### DISCUSSION TITLE
Autocompletion bug with multiple hyphens in options

### DISCUSSION THREAD

**Discussion by alexrosenfeld10** [fastapi/typer#965] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the Typer documentation, with the integrated search.
- [X] I already searched in Google "How to X in Typer" and didn't find any information.
- [X] I already read and followed all the tutorials in the docs and didn't find an answer.
- [X] I already checked if it is not related to Typer but to [Click](https://github.com/pallets/click).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import typer


app = typer.Typer(help="Testing 1234", no_args_is_help=True)

def autocomplete_test(incomplete: str):
    vals = ["foo-test-1234", "foo-anothertest-1234", "bar-test-1234"]
    return [val for val in vals if incomplete in val]


@app.command()
def test(
    name: str = typer.Option(None, "--name", "-n", help="Filter by name", autocompletion=autocomplete_test),
):
    typer.echo(f"foo {name}")
```

### Description

When the above code is run in a CLI and the input `my_cli test --name foo[TAB]` is entered into the shell (where `[TAB]` indicates the user pressing tab to autocomplete the `--name` argument), the following is the result of the autocompletion:

```
❯ my_cli test --name foo--1234
```

I have observed similar behaviors in slightly different scenarios, for example, if the autocomplete function looks like this instead:

```python
def autocomplete_test(incomplete: str):
    vals = ["foo-test-1234", "bar-test-1234"]
    return [val for val in vals if incomplete in val]
``` 

then a similar behavior happens when running `my_cli test --name [TAB]`:

```
❯ my_cli test --name -test-1234
```

Note - if the autocompletion options only have _one_ typo, instead of two or more, it works fine. As in, this has no issue:

```python
def autocomplete_test(incomplete: str):
    vals = [
        "foo-test",
        "foo-anothertest",
        "bar-test",
    ]
    return [val for val in vals if incomplete in val]
```

### Operating System

macOS

### Typer Version

0.12.5

### Python Version

3.12.3

**Comment by alexrosenfeld10** [comment]
@svlandeg @tiangolo this seems like a pretty clear bug, I took a look at the source code but didn't find any unit tests that trigger tab completion. If you can point me in the right direction I'll take a stab at fixing it

**Comment by svlandeg** [comment]
Hi @alexrosenfeld10, thanks for the report! 

You can find some autocompletion tests in `tests/test_completion`, perhaps you have a look whether you can contribute some unit tests that fail on `master`? And then we can start looking into fixing the bug 🙏

**Reply by alexrosenfeld10** [reply]
Hmm. I tried this and wasn't able to reproduce through the unit tests. I'll find some time to make a minimal reproduction project and share it here

**Reply by alexrosenfeld10** [reply]
@svlandeg this doesn't seem to reproduce easily via unit tests... however, I've reproduced it for you here: https://github.com/alexrosenfeld10/typer-autocomplete-bug

Thanks in advance for taking a look. I'll be responsive here or in the issue page in my repo if you need any help setting up or reproducing the problem.

**Reply by alexrosenfeld10** [reply]
@svlandeg let me know if there's anything else I can to do help debug this

**Reply by alexrosenfeld10** [reply]
Hi @svlandeg , did you get a chance to try out the reproduction project? If you can point me in the right direction I'll take a stab at fixing the issue.

**Comment by alexrosenfeld10** [comment]
@tiangolo possible this could get looked at? It's a pretty common case to have hyphenated autocomplete suggestions, and this totally breaks that

**Comment by alexrosenfeld10** [comment]
@tiangolo would be great if we could get some eyes on this, having autocomplete is one of the main things typer writes about in the homepage. Again I'm happy to help if you can point me in the right direction

**Comment by svlandeg** [comment]
Hi @alexrosenfeld10, I've finally had some time to look at this in detail. Thanks for the detailed report and repo to reproduce.

I've tried to write some unit tests to reproduce this issue, but I couldn't - they are all green on `master`: https://github.com/fastapi/typer/pull/1095

I've looked at the details of the examples you've provided. For instance:

```
vals = ["foo-test-1234", "foo-anothertest-1234", "bar-test-1234"]
```
and autocompleting with 
```
--name foo[TAB] 
```
leads to 
```
--name foo--1234
```
Which seems wrong, as `"foo--1234"` is not a correct string from the options. But, importantly, as I can see from your video from the repo you created to reproduce, the cursor in this case would still be _after_ `"foo-"` and _before_ `"-1234"`, right? Because in that sense the autocompletion is "correct": when starting with `"foo"` there are two options, both of which starting with `"foo-"` and ending in `"-1234"`. But you still need to fill in the correct middle part.

In this sense, I wonder whether maybe your title and focus on "two hyphens" is a red herring: I think what really matters here is that the beginning and end string are the same for two options. The same is true on your reproduction repo, where the options are 
```
"start-test-exmpl",
"start1234-other-exmpl",
"start1234-test-exmpl",
```

and the auto completion gives `"start--exmpl"` because that's the part of the string that overlaps between all three options 🤔 

So, one question I have is this: what happens when you remove the hyphens from the strings, and have e.g. options
```
vals = ["footest1234", "fooanothertest1234", "bartest1234"]
```
and autocomplete on
```
--name foo[TAB] 
```
does it lead to 
```
--name foo|1234
```
with `|` the meta mark for the cursor? Because if this is the case, then the behaviour you're seeing is not caused by the multiple hyphens, but by the string overlaps.

Then one other question: what happens if you hit TAB twice and not just once? Do you see the full options as strings?

Because on my system this does in fact work:

```
> typer disc_965.py run --name foo- [TAB TAB]
```

gives me as output:
```
foo-anothertest-1234  foo-test-1234
```
i.e. the two correct options are given in full. This is also what we see in the succeeding unit tests of #1095.

**Reply by alexrosenfeld10** [reply]
@svlandeg Thank you for checking back in on this! I appreciate the detailed reply. Responding in parts:

> in that sense the autocompletion is "correct"

I get what you mean, but don't think it's _functionally_ correct, as far as what the user would expect to happen. Hopefully we agree on that?

> So, one question I have is this: what happens when you remove the hyphens from the strings, and have e.g. options

I've just tested this and it works fine for me. In other words, it does _not_ lead to `--name foo|1234`, it leads to 

<img width="541" alt="Screenshot 2024-12-19 at 7 50 24 PM" src="https://github.com/user-attachments/assets/18a6a8cb-5db6-4809-b64a-af213d1315ed" />

Hitting tab multiple times toggles me through the various options as you'd expect.

**Reply by svlandeg** [reply]
Ok cool, that's good to know. So it looks like there's some interplay between the multiple hyphens and the string overlap... I'll continue to look into this.

**Reply by alexrosenfeld10** [reply]
@svlandeg just checking in on this, thanks in advance. if there's anything I can do to help let me know

**Reply by svlandeg** [reply]
Still looking into this. Quick update: I can now reproduce this locally (it seems to only happen with the `zsh` shell) and found that a single hyphen can already trigger this, e.g. with options
```
vals = ["fooQtest-1234", "fooQanothertest-1234", "barQtest-1234"]
```
and completing at `fooQ` gives `fooQ|-1234`. Note that it doesn't give `fooQ|test-1234` which is interesting. Somehow somewhere someone is spawning that end bit following the hyphen. I'll keep digging. Still can't reproduce with a unit test BTW.

**Reply by alexrosenfeld10** [reply]
Yeah, I concur it seems to be a zsh-specific problem.

**Reply by svlandeg** [reply, marked as answer]
So to come back to this observation I made earlier:

> ```
> vals = ["foo-test-1234", "foo-anothertest-1234", "bar-test-1234"]
> ```
> 
> and autocompleting with
> 
> ```
> --name foo[TAB] 
> ```
> 
> leads to
> 
> ```
> --name foo--1234
> ```
> 
> Which seems wrong, as `"foo--1234"` is not a correct string from the options. But, importantly, as I can see from your video from the repo you created to reproduce, the cursor in this case would still be _after_ `"foo-"` and _before_ `"-1234"`, right? Because in that sense the autocompletion is "correct": when starting with `"foo"` there are two options, both of which starting with `"foo-"` and ending in `"-1234"`. But you still need to fill in the correct middle part.

I am not that familiar with `zsh`, but the behaviour I describe above might actually be intentional. For instance, [this post](https://unix.stackexchange.com/questions/259287/how-can-i-get-zshs-completion-working-in-the-middle-of-the-filename) lists it as correct behaviour in its first part:

![image](https://github.com/user-attachments/assets/82bda121-57fb-459f-8d07-a76b0b5b4976)

So basically I'm starting to wonder whether this is really just a `zsh` setting, and nothing to do with Typer really.

More reading: https://zsh.sourceforge.io/FAQ/zshfaq04.html#l50 & https://superuser.com/a/603017

**Reply by svlandeg** [reply]
For instance, when I put this in my .zshc profile:
```
zstyle ':completion:*' completer _list _expand _complete _approximate _match
```

then the value isn't being completed with the cursor in the middle, instead it just immediately shows me the two correct options:
```
typer disc_965.py run --name f
foo-anothertest-1234  foo-test-1234
```

**Reply by alexrosenfeld10** [reply]
@svlandeg that works for me as well but has another downside - when there's only one option, it no longer auto-fills, instead it prints the "list" of options as if there were multiple, even though there is only one

**Reply by alexrosenfeld10** [reply]
think you're right though, generally this doesn't seem like a typer bug, more of a zsh / omz issue

**Reply by alexrosenfeld10** [reply]
This: https://github.com/ohmyzsh/ohmyzsh/issues/1398#issuecomment-255596509 seems to be the silver bullet for me

**Reply by alexrosenfeld10** [reply]
Ugh, the above change breaks other types of tab completion. For example:
```shell
my-tool some-arg --<tab>
```

results in 
```
my-tool some-arg --he
```

even though the options are `--help` and `--lightweight`. 

Back to the drawing board.... still not a typer issue though. Mostly writing this in case someone finds it in future
